Support for pkcs1

cwperks opened the pull request:
Signed-off-by: Craig Perkins <cwperx@amazon.com>

Motivation:

Motivated by a stale [PR](https://github.com/netty/netty/pull/7451) that was closed. This change adds support for keys in the PKCS#1 format. Currently netty only supports PKCS#8 keys.

Modification:

This change introduces a class called `BouncyCastlePemReader` which is only used if BouncyCastle is available on the classpath and uses BouncyCastle's PEMParser to parse the private keys. See list of supported types [here](https://www.bouncycastle.org/docs/pkixdocs1.5on/org/bouncycastle/openssl/PEMParser.html). 

Tests are added in `SSLContextTest` including tests with PKCS#8 keys (encrypted + unencrypted) to show these working with BouncyCastle.

Result:

Fixes [#7323](https://github.com/netty/netty/issues/7323)

normanmaurer commented on handler/src/main/java/io/netty/handler/ssl/SslContext.java:
```diff
@@ -1120,6 +1121,15 @@ protected static PrivateKey toPrivateKey(File keyFile, String keyPassword) throw
+
+        // try BC first, if this fail fallback to original key extraction process
+        if (BouncyCastlePemReader.isAvailable()) {
+            PrivateKey pk;
+            if ((pk = BouncyCastlePemReader.getPrivateKey(keyFile, keyPassword)) != null) {
+                return pk;
+            }
```
```suggestion
            PrivateKey pk = BouncyCastlePemReader.getPrivateKey(keyFile, keyPassword);
            if (pk != null) {
                return pk;
            }
```

normanmaurer commented on handler/src/main/java/io/netty/handler/ssl/SslContext.java:
```diff
@@ -1131,6 +1141,15 @@ protected static PrivateKey toPrivateKey(InputStream keyInputStream, String keyP
         }
+
+        // try BC first, if this fail fallback to original key extraction process
+        if (BouncyCastlePemReader.isAvailable()) {
+            PrivateKey pk;
+            if ((pk = BouncyCastlePemReader.getPrivateKey(keyInputStream, keyPassword)) != null) {
+                return pk;
+            }
+        }
+
```
```suggestion
        if (BouncyCastlePemReader.isAvailable()) {
            PrivateKey pk = BouncyCastlePemReader.getPrivateKey(keyInputStream, keyPassword);
            if (pk != null) {
                return pk;
            }
        }
```

peternied replied:
+1 I am not a fan of inline assignments as they make the code harder to read

normanmaurer commented on handler/src/main/java/io/netty/handler/ssl/util/BouncyCastlePemReader.java:
```diff
@@ -0,0 +1,214 @@
+                    logger.debug("Bouncy Castle provider available");
+                    attemptedLoading = true;
+                    return Boolean.TRUE;
+                } catch (Exception e) {
```
```suggestion
                } catch (Throwable e) {
```

normanmaurer commented on handler/src/main/java/io/netty/handler/ssl/util/BouncyCastlePemReader.java:
```diff
@@ -0,0 +1,214 @@
+                    bcProvider = bcProviderClass.getConstructor().newInstance();
+                    logger.debug("Bouncy Castle provider available");
+                    attemptedLoading = true;
+                    return Boolean.TRUE;
```
Seems like the return value is never used ?

cwperks replied:
Updated this to `PrivilegedAction<Void>` since the return value is not used.

normanmaurer commented on handler/src/main/java/io/netty/handler/ssl/util/BouncyCastlePemReader.java:
```diff
@@ -0,0 +1,214 @@
+        if (!isAvailable()) {
+            if (logger.isDebugEnabled()) {
+                logger.debug("Bouncy castle provider is unavailable.", unavailabilityCause());
+            }
```
return early ?

normanmaurer commented on handler/src/main/java/io/netty/handler/ssl/util/BouncyCastlePemReader.java:
```diff
@@ -0,0 +1,214 @@
+            if (logger.isDebugEnabled()) {
+                logger.debug("Bouncy castle provider is unavailable.", unavailabilityCause());
+            }
+        }
```
return early ?

normanmaurer commented on handler/src/main/java/io/netty/handler/ssl/util/BouncyCastlePemReader.java:
```diff
@@ -0,0 +1,214 @@
+            if (logger.isDebugEnabled()) {
+                logger.debug("Bouncy castle provider is unavailable.", unavailabilityCause());
+            }
+        }
```
return early ?

peternied reviewed (approved):
I agree with the comments about readability, but its a major improvement so other libraries/products don't have to do this themselves.

cwperks commented:
@normanmaurer Thank you for the review. I have addressed the code review comments and moved this block:

```
if (!isAvailable()) {
    if (logger.isDebugEnabled()) {
        logger.debug("Bouncy castle provider is unavailable.", unavailabilityCause());
    }
    return null;
}
```

into the `public static PrivateKey BouncyCastlePemReader.getPrivateKey(...)` methods and exit early if BC is unavailable.

normanmaurer commented on handler/src/main/java/io/netty/handler/ssl/util/BouncyCastlePemReader.java:
```diff
@@ -0,0 +1,211 @@
+import org.bouncycastle.pkcs.PKCS8EncryptedPrivateKeyInfo;
+import org.bouncycastle.pkcs.PKCSException;
+
+public final class BouncyCastlePemReader {
```
I wonder we should move this to the ssl package and make it package-private. WDYT ?

cwperks replied:
Moved this to `io.netty.handler.ssl` and removed the `public` access modifier. This is only used in `SslContext` which is the same package so it does not need to be `public`.

normanmaurer commented:
@cwperks thanks a lot!
